### DISCUSSION TITLE
Using igl::tetrahedralize with an internal cavity?

### DISCUSSION THREAD

**Discussion by TheBEllis** [libigl/libigl#2056] [Q&A] [answered]
Hi libIGL team. I was wondering whether the tetGen based igl::tetrahedralize can be used to create a tet-mesh with an internal cavity. I have a simple cube surface mesh made up of tri elements, and I want to tetrahedralize the internal volume of this cube, but with a cavity in the middle of it. This cavity is defined by another surface tri mesh, that can be of arbitrary shape but does not overlap with the cube surface. Essentially I'm looking for a 3D version of the 2D example in the tutorial (https://libigl.github.io/images/604_Triangle.png) for the triangulation of closed polygons. I was just wondering whether you know if this is possible using igl::tetrahedralize. I hope I have explained the problem okay, any help is really appreciated :)

**Comment by alecjacobson** [comment, marked as answer]
Indeed it's possible! see the "seed points inside holes" at https://github.com/libigl/libigl/blob/3370a3e9caab4708e02f00a5ae34f6aaba27a428/include/igl/copyleft/tetgen/tetrahedralize.h#L130

Alternatively, you could mesh the whole shape with all the faces as input (e.g., including the `-c` flag) and then extract tets as needed according to their winding number (see `winding_number`).

**Reply by TheBEllis** [reply]
Thank you, I will take a look!

**Reply by TheBEllis** [reply]
Hi Alec, I was wondering if you could offer some guidance on the use of region attributes. I don't currently need to make use of this as I don't have any spatially varying properties that I need defined at the moment, is it suitable just to pass in an empty eigen matrix like so?
`Eigen::MatrixXd R;` , or should I define one region for the entire mesh. Also for the seed points, should there be one point set for each closed hole or cavity, or would multiple points inside the same cavity offer some advantage?

**Reply by alecjacobson** [reply]
yes, empty should work for `R`.

There definitely needs to be at least one point per cavity. I haven't tried multiple. I'd hope it'd work, but I'm not sure.

**Reply by TheBEllis** [reply]
Thanks, I didn't want to have to go and set some regions arbitrarily:) For anyone else trying to do this, I have found using .OFF mesh files to work better than .mesh files. Doing this allowed the tetrahedralisation process to work, whereas with .mesh files I could only produce the convex hull without getting any errors. I think this may be something to do with how my indices were numbered in the .mesh files. I used meshio to convert from exodus to .mesh, and printing out my indices after reading in the .mesh file using libIGL, they were non-zero based. So this may well be a meshio problem, probably completely unrelated to the Medit reader in libIGL, but may be helpful to know for anyone else having issues!
